src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHelpers.Unix.cs

```diff
@@ -0,0 +1,150 @@
+namespace System.Formats.Tar
+{
+    internal static partial class TarHelpers
+    {
+        private static readonly Lazy<UnixFileMode> s_umask = new Lazy<UnixFileMode>(DetermineUMask);
+
+        private static UnixFileMode DetermineUMask()
+        {
+            // To determine the umask, we'll create a file with full permissions and see
+            // what gets filtered out.
+            // note: only the owner of a file, and root can change file permissions.
+
+            const UnixFileMode OwnershipPermissions =
+                UnixFileMode.UserRead | UnixFileMode.UserWrite | UnixFileMode.UserExecute |
+                UnixFileMode.GroupRead | UnixFileMode.GroupWrite | UnixFileMode.GroupExecute |
+                UnixFileMode.OtherRead | UnixFileMode.OtherWrite |  UnixFileMode.OtherExecute;
+
+            string filename = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName());
+            FileStreamOptions options = new()
+            {
+                Mode = FileMode.CreateNew,
+                UnixCreateMode = OwnershipPermissions,
+                Options = FileOptions.DeleteOnClose,
+                Access = FileAccess.Write,
+                BufferSize = 0
+            };
+            using var fs = new FileStream(filename, options);
+            UnixFileMode actual = File.GetUnixFileMode(fs.SafeFileHandle);
+
+            return OwnershipPermissions & ~actual;
+        }
+
+        private sealed class ReverseStringComparer : IComparer<string>
+        {
+            public int Compare (string? x, string? y)
+                => StringComparer.Ordinal.Compare(y, x);
+        }
+
+        private static readonly ReverseStringComparer s_reverseStringComparer = new();
+
+        private static UnixFileMode UMask => s_umask.Value;
+
+        /*
+            Tar files are usually ordered: parent directories come before their child entries.
+
+            They may be unordered. In that case we need to create parent directories before
+            we know the proper permissions for these directories.
+
+            We create these directories with restrictive permissions. If we encounter an entry for
+            the directory later, we store the mode to apply it later.
+
+            If the archive doesn't have an entry for the parent directory, we use the default mask.
+
+            The pending modes to be applied are tracked through a reverse-sorted dictionary.
+            The reverse order is needed to apply permissions to children before their parent.
+            Otherwise we may apply a restrictive mask to the parent, that prevents us from
+            changing a child.
+        */
+
+        internal static SortedDictionary<string, UnixFileMode>? CreatePendingModesDictionary()
+            => new SortedDictionary<string, UnixFileMode>(s_reverseStringComparer);
+
+        internal static void CreateDirectory(string fullPath, UnixFileMode? mode, bool overwriteMetadata, SortedDictionary<string, UnixFileMode>? pendingModes)
```

Comment:
Instead of passing a [nullable value type](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/nullable-value-types), why not move the `mode` parameter to the last position and set a default value of `UnixFileMode.None`?

Then the lines below, instead of checking `mode.HasValue`, they could check for `mode is not UnixFileMode.None`.

The reason I'm suggesting this is because I suspect that the boolean logical operations below, which are consuming the `mode` directly, are not really operating on the `mode.Value`, so the result might not be the expected one (Please correct me if I'm wrong here).

Reply:
```cs
        internal static void CreateDirectory(
            string fullPath,
            bool overwriteMetadata,
            SortedDictionary<string, UnixFileMode> pendingModes,
            UnixFileMode mode = UnixFileMode.None)
```

Reply:
> Instead of passing a [nullable value type](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/nullable-value-types), why not move the mode parameter to the last position and set a default value of UnixFileMode.None?

`null` is used when we need to create a directory but we don't know the mode from the archive (yet).

A non-`null` mode (including `UnixFileMode.None`) means it is the mode that was stored in the archive for `fullPath`.